tokenizer/src/lib.rs

```diff
@@ -485,3 +491,53 @@ impl<'a> TokenStream<'a> {
         Token { start, kind }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use crate::{Token, TokenKind, TokenStream};
+
+    /// A wrapper over [`TokenStream`] that iterates over tokens and returns
+    /// them as pairs of the token kind and the subslice of the buffer that
+    /// corresponds to the token.
+    struct TokenAndSliceIterator<'a> {
+        stream: TokenStream<'a>,
+        slice: &'a [u8],
+        current_token: Option<Token>,
+    }
+    impl<'a> TokenAndSliceIterator<'a> {
+        pub fn new(buffer: &'a [u8]) -> TokenAndSliceIterator<'a> {
+            TokenAndSliceIterator {
+                stream: TokenStream::new(buffer),
+                slice: buffer,
+                current_token: None,
+            }
+        }
+    }
+    impl<'a> Iterator for TokenAndSliceIterator<'a> {
+        type Item = (TokenKind, &'a [u8]);
+
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.current_token.is_none() {
+                self.current_token = Some(self.stream.next());
+            }
```

Comment:
It's a small issue, but if you could match the `current_token` and reassigning, then take the resulting token as a local, the later unwraps would be unnecessary.
```suggestion
            let current_token = match self.current_token {
                Some(token) => token,
                None => {
                    let token = self.stream.next();
                    self.current_token = Some(token);
                    token
            };
```